services/stip_relayer/main.go

```diff
@@ -0,0 +1,434 @@
+}
+
+type CustomTime struct {
+	time.Time
+}
+
+const ctLayout = "2006-01-02 15:04:05.000 MST"
+
+func (ct *CustomTime) UnmarshalJSON(b []byte) error {
+	s := strings.Trim(string(b), "\"")
+	if s == "null" {
+		return nil
+	}
+	t, err := time.Parse(ctLayout, s)
+	if err != nil {
+		return err
+	}
+	ct.Time = t
+	return nil
+}
+
+func (s *STIPRelayer) Run(ctx context.Context) error {
+	// Create a cancellable context
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel() // ensure cancel is called to clean up resources
+
+	g, ctx := errgroup.WithContext(ctx)
+
+	// Start the submitter goroutine
+	g.Go(func() error {
+		return s.StartSubmitter(ctx)
+	})
+
+	// Start the ticker goroutine for requesting and storing execution results
+	g.Go(func() error {
+		return s.RequestAndStoreResults(ctx)
+	})
+
+	// Start the goroutine for querying, rebating/relaying, and updating results
+	g.Go(func() error {
+		return s.QueryRebateAndUpdate(ctx)
+	})
+
+	// Wait for all goroutines to finish
+	if err := g.Wait(); err != nil {
+		return err // handle the error from goroutines
+	}
+
+	return nil
+}
+
+// startSubmitter handles the initialization of the submitter
+func (s *STIPRelayer) StartSubmitter(ctx context.Context) error {
+	err := s.submittter.Start(ctx)
+	if err != nil {
+		fmt.Printf("could not start submitter: %v", err)
+		return nil // return nil to keep other goroutines running
+	}
+	return nil
+}
+
+// requestAndStoreResults handles the continuous request of new execution results and storing them in the database
+func (s *STIPRelayer) RequestAndStoreResults(ctx context.Context) error {
+	ticker := time.NewTicker(5 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err() // exit if context is cancelled
+		case <-ticker.C:
+			if err := s.ProcessExecutionResults(ctx); err != nil {
+				// Log the error and decide whether to continue based on the error
+				fmt.Printf("Error processing execution results: %v", err)
+				// Optionally, you can return the error to stop the goroutine
+				// return err
+			}
+		}
+	}
+}
+
+// processExecutionResults encapsulates the logic for requesting and storing execution results
+func (s *STIPRelayer) ProcessExecutionResults(ctx context.Context) error {
+	resp, err := ExecuteDuneQuery()
+	if err != nil {
+		return fmt.Errorf("failed to execute Dune query: %v", err)
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return fmt.Errorf("failed to read response body: %v", err)
+	}
+
+	var result map[string]string
+	err = json.Unmarshal(body, &result)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal response body: %v", err)
+	}
+
+	executionID, ok := result["execution_id"]
+	if !ok {
+		return fmt.Errorf("no execution_id found in response")
+	}
+
+	time.Sleep(20 * time.Second) // Consider replacing this with a more robust solution
+
+	executionResults, err := GetExecutionResults(executionID)
```

Comment:
Eventually may want to wrap this in a `backoff.Retry()` then you don't have to have the hard-coded sleep

Reply:
resolved in https://github.com/synapsecns/sanguine/pull/1872/commits/64e60cbf5d52ccf282aef9eaded8107eefc0f02f